src/graphnet/models/flows/spline_blocks.py

```diff
@@ -0,0 +1,385 @@
+"""Utility classes for SplineBlocks for NormalizingFlows in GraphNeT."""
+
+from typing import Tuple, List
+from abc import abstractmethod
+import torch
+
+from graphnet.models import Model
+
+
+class RationalQuadraticSpline(Model):
+    """Implementation of https://arxiv.org/pdf/1906.04032.pdf, page 4."""
+
+    def __init__(self, n_knots: int, b: int):
+        """Construct `RationalQuadraticSpline`.
+
+        Args:
+            n_knots: Number of knots per spline.
+            b: Bounding parameter. Input is assumed to be in [-b,b].
+        """
+        super(RationalQuadraticSpline, self).__init__()
+        self.n_knots = n_knots
+        self.b = b
+        self._softmax = torch.nn.Softmax(dim=1)
+        self._softplus = torch.nn.Softplus()
+        self._eps = 0
+
+    def forward_spline(
+        self,
+        x: torch.Tensor,
+        spline_parameters: torch.Tensor,
+    ) -> torch.Tensor:
+        """Forward spline call.
+
+        Transform a dimension `x` from input distribution into dimension `y` in
+        latent distribution.
+        """
+        # Checks
+        assert ~torch.any(
+            ~(x.lt(self.b) & x.gt(-self.b)).bool()
+        ), f"At least one sample in `x` is out of bounds of [{-self.b}, {self.b}]: {torch.max(x, dim =1)[0]}"
+
+        # Transform knot bins to coordinates and pad as described in paper
+        knot_x, knot_y, d = self._partition_spline_parameters(
+            spline_parameters=spline_parameters
+        )
+        # Calculate knot index `k`, s and epsilon.
+        k = self._find_spline_idx(knot_x, x)
+        s = self._calculate_s(knot_y, knot_x, k)
+        epsilon = self._calculate_epsilon(knot_x, x, k)
+
+        # Evaluate Spline & get Jacobian
+        y = self._evaluate_spline(s, epsilon, knot_y, d, k)
+        jacobian = self._calculate_jacobian(s, d, k, epsilon)
+        return y.reshape(-1, 1).squeeze(1), jacobian.squeeze(1)
+
+    def inverse_spline(
+        self,
+        y: torch.Tensor,
+        spline_parameters: torch.Tensor,
+    ) -> torch.Tensor:
+        """Inverse spline call.
+
+        Transform of dimension `y` from internal distribution into dimension
+        `x` from input distribution.
+        """
+        # Checks
+        assert ~torch.any(
+            ~(y.lt(self.b) & y.gt(-self.b)).bool()
+        ), f"At least one sample in `x` is out of bounds of [{-self.b}, {self.b}]."
+
+        # Transform knot bins to coordinates and pad as described in paper
+        knot_x, knot_y, d = self._partition_spline_parameters(
+            spline_parameters=spline_parameters
+        )
+
+        # Calculate knot index `k` and s
+        k = self._find_spline_idx(knot_y, y)
+        assert (
+            max(k) + 1 <= knot_y.shape[1]
+        ), f"""{knot_y.shape} vs. {max(k) + 1}"""
```

Comment:
I think a more descriptive error here would help.

Reply:
This assert statement has been removed; the one above it effectively checks the same. Thanks for spotting it!